### PR TITLE
pgo: add support for clang's Profile Guided Optimization (PGO)

### PR DESCRIPTION
**Pull request by gwelymernans**
Profile-Guided Optimization (PGO) is clang's version of gcc's FDO. To
generate a profile, the kernel is instrumented with PGO counters. A
representative workload is then ran. The raw profile data can be
collected from /sys/kernel/debug/pgo/profraw.

The raw profile data must be processed by "llvm-profdata" before it can
be used during recompilation:

  $ cp /sys/kernel/debug/pgo/profraw vmlinux.profraw
  $ llvm-profdata merge -output vmlinux.profdata vmlinux.profraw

Multiple raw profiles may be merged into a single profile.

The data can be used either by the compiler if LTO isn't enabled:

    ... -fprofile-use=vmlinux.profdata ...

or by LLD if LTO is enabled:

    ... -lto-cs-profile-file=vmlinux.profdata ...

### CODE REVIEW

**Comment by gwelymernans** [comment]
There still needs to be a `Documentation/` entry still. But this is the basic bits.

File: arch/x86/Kconfig [hunk 1]
```diff
@@ -93,8 +93,9 @@ config X86
 	select ARCH_SUPPORTS_ACPI
 	select ARCH_SUPPORTS_ATOMIC_RMW
 	select ARCH_SUPPORTS_DEBUG_PAGEALLOC
-	select ARCH_SUPPORTS_NUMA_BALANCING	if X86_64
 	select ARCH_SUPPORTS_KMAP_LOCAL_FORCE_MAP	if NR_CPUS <= 4096
+	select ARCH_SUPPORTS_NUMA_BALANCING	if X86_64
```
Comment (samitolvanen) [inline arch/x86/Kconfig:97]: No need to move the `ARCH_SUPPORTS_NUMA_BALANCING` line.

File: kernel/pgo/Kconfig [hunk 2]
```diff
@@ -0,0 +1,33 @@
+# SPDX-License-Identifier: GPL-2.0-only
+menu "Profile Guided Optimization (PGO) (EXPERIMENTAL)"
+
+config ARCH_SUPPORTS_PGO_CLANG
+	def_bool n
```
Comment (samitolvanen) [inline kernel/pgo/Kconfig:5]: I'm not sure if `def_bool` is needed here as `bool` should default to no, but a help text might be useful.

File: kernel/pgo/instrument.c [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+	/* Lock when updating the value node structure */
+	spin_lock_irqsave(&node_lock, flags);
+
+	curr = allocate_node(p, index, target_value);
+	if (curr) {
+		curr->value = target_value;
+		curr->count++;
+
+		if (!counters[index])
+			cmpxchg(&counters[index], NULL, curr);
```
Comment (samitolvanen) [inline kernel/pgo/instrument.c:117]: All the locking code should be double checked for sanity. Do we need the `serializing` atomic? And re-reading this now, I'm not entirely sure why we're using `cmpxchg` here, and why aren't we using it in a loop if we care about performing the assignment? Does this leak `curr` if the assignment fails? At the very least, this should be commented to explain what's going on.

File: arch/x86/Kconfig (hunk 1 above)
Comment (gwelymernans) [reply arch/x86/Kconfig:97]: It's not alphabetized!!11!1!ELEVEN!! *twitch*

Okay, I'll revert that bit.

File: kernel/pgo/Kconfig (hunk 2 above)
Comment (gwelymernans) [reply kernel/pgo/Kconfig:5]: I was cribbing off of `kernel/gcov/Kconfig`. It seems like these kind of config options don't have help messages associated with them. I'm not entirely sure why.

File: kernel/pgo/instrument.c (hunk 3 above)
Comment (gwelymernans) [reply kernel/pgo/instrument.c:117]: But you wrote it! :-P

I'll double check the code.

**Comment by nickdesaulniers** [comment]
can we merge this into a branch other than master? @nathanchance 's cron jobs to push to master from mainline are likely failing as a result of the merge.  Also, I'd like to take the time to help test and review this, but after the holidays.

**Comment by nickdesaulniers** [comment]
oh, I see what happened. No worries, @nathanchance should just force push the latest copy of mainline.

**Comment by gwelymernans** [comment]
Sorry about this. I hate Git for reasons. :-(

**Comment by nathanchance** [comment]
I handled it last night when I saw what happened.

It might be worth using a separate branch in this repository and commenting on the issue asking for review, rather than using pull requests because we never intend to merge into this tree. Although I guess reviewing a PR is a little cleaner than adding individual comments... no strong preference, merely throwing ideas out.

**Comment by nickdesaulniers** [comment]
When you issue a PR against a github repo, you can choose the destination branch.  Though we don't intend to merge to master/main, I'm not against storing WIP patches on branches. It makes sharing patches trivial.  In github's UI for PRs, you can select a branch that's not master for a PR. PR is required for the code review interface, though if you just push to branch, folks can leave comments and you can force push to the branch with updates.

**Comment by nathanchance** [comment]
I could add a "for-review" branch that is kept in sync with master that people can target for merges.

**Pull request merged**